rewriter/ChainedSig.cc

```diff
@@ -0,0 +1,178 @@
+        while (sigBlock) {
+            if (sigBlock->fun == fun) {
+                if (auto e = ctx.beginError(sigBlock->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                    e.setHeader("Duplicate invocation of `{}` in signature declaration", fun.toString(ctx));
+                }
+            }
+
+            sigBlock = ast::cast_tree<ast::Send>(sigBlock->recv);
+        }
+    }
+
+    bool invalidChainedStatement(core::MutableContext ctx, ast::Send *send) {
+        if (send->fun == core::Names::params() || send->fun == core::Names::returns() ||
+            send->fun == core::Names::void_() || send->fun == core::Names::bind() ||
+            send->fun == core::Names::checked() || send->fun == core::Names::onFailure() ||
+            send->fun == core::Names::typeParameters()) {
+            if (auto e = ctx.beginError(send->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Cannot use `{}` outside of a sig block", send->fun.toString(ctx));
+            }
+
+            return true;
+        }
+
+        return false;
+    }
+
+    ast::ExpressionPtr buildReplacement(core::MutableContext ctx, ast::Send *send, ast::Send *blockBody) {
+        ast::Send::ARGS_store args;
+        auto treeCopy = send->deepCopy();
+        auto sendCopy = ast::cast_tree<ast::Send>(treeCopy);
+
+        // Go through each part of the chained sig and make sure it's not duplicated inside the block
+        // E.g.: `sig.abstract { abstract.void }`
+        while (sendCopy && sendCopy->fun != core::Names::sig()) {
+            checkDuplicates(ctx, blockBody, sendCopy->fun);
+
+            sendCopy = ast::cast_tree<ast::Send>(sendCopy->recv);
+        }
+
+        sendCopy = ast::cast_tree<ast::Send>(treeCopy);
+        bool isFinal = false;
+
+        // Create a new receiver for the block to move the statements chained on sig inside. Also, finds out whether the
+        // signature is final or not
+        ast::ExpressionPtr newBlockReceiver = std::move(blockBody->recv);
+        do {
+            if (sendCopy->fun != core::Names::final_()) {
+                if (sendCopy->hasBlock()) {
+                    // Drop the block argument before moving the arguments.
+                    sendCopy->setBlock(nullptr);
+                }
+                newBlockReceiver = sendCopy->withNewBody(sendCopy->loc, std::move(newBlockReceiver), sendCopy->fun);
+            } else {
+                isFinal = true;
+            }
+
+            sendCopy = ast::cast_tree<ast::Send>(sendCopy->recv);
+
+            // If a previous receiver has a block, then we need to add an error to prevent signatures like this:
+            // `sig.final {}.override{}
+            if (sendCopy->hasBlock()) {
+                if (auto e = ctx.beginError(send->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                    e.setHeader("Cannot add more signature statements after the declaration block");
+                }
```

Comment:
Highlighting the entire send here doesn't give the user a clear picture of where the error is; we should be able to highlight the problematic part of the send directly, and (maybe) provide information about the original declaration block.